dronegen/container_images.go

```diff
@@ -0,0 +1,863 @@
+// sed -i '' 's/type\: kubernetes/type\: docker/' .drone.yml && sed -i '' 's/type\: exec/type\: docker/' .drone.yml
+// # Drone has a bug where "workspace" is appended to "/drone/src". This fixes that by updating references
+// sed -i '' 's~/go/~/drone/src/go/~g' .drone.yml
+// # Pull the current branch instead of v10
+// sed -i '' "s~git checkout -qf \"\$(cat '/go/vars/full-version/v10')\"~git checkout -qf \"${DRONE_SOURCE_BRANCH}\"~" .drone.yml
+// # `drone exec` does not properly map the workspace path. This creates a volume to be shared between steps
+// #  at the correct path
+// DOCKER_VOLUME_NAME="go"
+// docker volume create "$DOCKER_VOLUME_NAME"
+// drone exec --trusted --pipeline teleport-container-images-current-version-cron --clone=false --volume "${DOCKER_VOLUME_NAME}:/go"
+// # Cleanup
+// docker volume rm "$DOCKER_VOLUME_NAME"
+
+import (
+	"fmt"
+	"math"
+	"path"
+	"strings"
+)
+
+// If you are working on a PR/testing changes to this file you should configure the following for Drone testing:
+// 1. Publish the branch you're working on
+// 2. Set `PRBranch` to the name of the branch in (1)
+// 3. Set `ConfigureForPRTestingOnly` to true
+// 4. Create a public and private ECR, Quay repos for "teleport", "teleport-ent", "teleport-operator", "teleport-lab"
+// 5. Set `TestingQuayRegistryOrg` and `TestingECRRegistryOrg` to the org name(s) used in (4)
+// 6. Set the `ECRTestingDomain` to the domain used for the private ECR repos
+// 7. Create two separate IAM users, each with full access to either the public ECR repo OR the private ECR repo
+// 8. Create a Quay "robot account" with write permissions for the created Quay repos
+// 9. Set the Drone secrets for the secret names listed in "GetContainerRepos" to the credentials in (7, 8), prefixed by the value of `TestingSecretPrefix`
+//
+// On each commit, after running `make dronegen``, run the following commands and resign the file:
+// # Pull the current branch instead of v10 so the appropriate dockerfile gets loaded
+// sed -i '' "s~git checkout -qf \"\$(cat '/go/vars/full-version/v10')\"~git checkout -qf \"${DRONE_SOURCE_BRANCH}\"~" .drone.yml
+//
+// When finishing up your PR check the following:
+// * The testing secrets added to Drone have been removed
+// * `ConfigureForPRTestingOnly` has been set to false, and `make dronegen` has been reran afterwords
+
+const (
+	ConfigureForPRTestingOnly bool   = true
+	TestingSecretPrefix       string = "TEST_"
+	TestingQuayRegistryOrg    string = "fred_heinecke"
+	TestingECRRegistryOrg     string = "u8j2q1d9"
+	TestingEcrRegion          string = "us-east-2"
+	PRBranch                  string = "fred/arm-container-images"
+	ECRTestingDomain          string = "278576220453.dkr.ecr.us-east-2.amazonaws.com"
+)
+
+const (
+	ProductionRegistryOrg string = "gravitational"
+	PublicEcrRegion       string = "us-east-1"
+	StagingEcrRegion      string = "us-west-2"
+
+	LocalRegistry string = "drone-docker-registry:5000"
```

Comment:
What is the local registry used for?

Reply:
`docker buildx` will not use the normal `docker image` local registry with the `docker-container` driver. Running a registry in the context of a pipeline allows layers/images to be shared around steps without exporting everything as tarballs and passing the files around multiple steps via the workspace dir. Strictly speaking this can be removed/replaced but it makes the build, tag, and push steps much more messy.

Reply:
Just to clarify, `docker buildx` doesn't use the local `docker image ls`  when using the `docker-container` driver. And this driver is used because of docker-in-docker?

Reply:
For the sake of educating myself on this issue I looked into the problem you are facing. Looks like `docker-container` driver does support the local registry but only when you are building for a single platform at a time. Running something like `--platform linux/amd64,linux/arm64` doesn't work as you mention. 

I have two conflicting comments I've already made. The first is not using docker registry. This can be done because you could possibly switch back to using `--load` instead of `--push` but you'll have to individually run buildx build for each platform. 

Alternatively I mentioned leveraging `--platform` so that you can build them all at once in one command. This doesn't work with `--load` so you'll have to leverage `--push` with the registry like you currently do. 

You currently leverage the registry and individual platform builds. I suggest you switch to one or the other. I tested this locally, but I could be overlooking something though so let me know. 

Note: https://docs.docker.com/desktop/containerd/ Containerd image store beta fixes this issue entirely allowing you to leverage all platforms in one command and still not have to use the registry solution which seems to be the optimal path. This should come out soon.

Reply:
> Just to clarify, docker buildx doesn't use the local docker image ls when using the docker-container driver. And this driver is used because of docker-in-docker?

Yes that is correct, it is due entirely to the driver and is unrelated to DinD. One of the biggest places this causes and issue is when building teleport lab. Teleport lab depends on the base `teleport` image which is also built in this pipeline. Without using a registry there is no clean way to provide a multiarch `BASE` to teleport operator's dockerfile.

I can absolutely switch to using `--platform` with multiple arch/targets. The reason why I have not here is because (IMO) splitting the steps up makes it easier to diagnose and debug an issue when something goes wrong during a promotion. For example, if the `teleport-ent:v10-arm` build fails due to a change with what cross compiler package we're using, it's much easier to track down that issue than if `teleport-env:v10` targeting `arm, arm64, amd64` fails.

Reply:
Sounds good to me! In that case it seems like you can remove the need for the registry then as building individually doesn't cause issues with loading into `docker image`.